[llvm/CAS] Reduce uses of `LLVM_ENABLE_ONDISK_CAS` macro

akyrtzi opened the pull request:
`LLVM_ENABLE_ONDISK_CAS` in `llvm/CAS` headers is not a good idea because any file that includes them needs to get the macro definition. Reduce use of the macro and use it only for:

* Removing the functionality from the high-level CAS creation functions.
* Removing the functionality from tests.

Essentially we should treat `LLVM_ENABLE_ONDISK_CAS` as a temporary macro until Windows is fully supported.

finagolfin committed:
[CMake] Only enable LLVM_ENABLE_ONDISK_CAS_default for 64-bit or larger architectures

(cherry picked from commit 3fea15d271ae286b0a9644f637b28c40c8e72c03)

akyrtzi committed:
[llvm/CAS] Reduce uses of `LLVM_ENABLE_ONDISK_CAS` macro

`LLVM_ENABLE_ONDISK_CAS` in `llvm/CAS` headers is not a good idea because any file that includes
them needs to get the macro definition. Reduce use of the macro and use it only for:

* Removing the functionality from the high-level CAS creation functions.
* Removing the functionality from tests.

Essentially we should treat `LLVM_ENABLE_ONDISK_CAS` as a temporary macro until Windows is fully supported.

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
I've verified that Windows compiles successfully after these changes using swift PR testing.

cachemeifyoucan commented on llvm/lib/CAS/OnDiskHashMappedTrie.cpp:
```diff
@@ -18,8 +18,6 @@
-
 using namespace llvm;
 using namespace llvm::cas;
 
```
The main goal of the macro/cmake configuration is to guard the static_assert below, otherwise it is going to fail to build on 32 bit arch. If the goal is thoroughly get the code to build, everything probably compiles fine on all platforms with those assertion removed.

cachemeifyoucan replied:
Github won't let me comment on the line that is not in the diff. It is basically the line after here but it is not shown in snippet.

akyrtzi replied:
> The main goal of the macro/cmake configuration is to guard the static_assert below, otherwise it is going to fail to build on 32 bit arch.

That's not what controls the configuration though, in CMake it is
```
if(UNIX)
  set(LLVM_ENABLE_ONDISK_CAS_default ON)
else()
  set(LLVM_ENABLE_ONDISK_CAS_default OFF)
endif()
```
which indicates it's only disabled by default on windows, the architecture is not checked.

cachemeifyoucan replied:
Whether it is enabled or not by default is not the point, the point is that it is a switch you can turn off if it doesn't work on your platform.

So technically, this PR adds on swift branch from external developer that wasn't landed on other branches: https://github.com/apple/llvm-project/pull/6098

akyrtzi replied:
Thanks for pointing out that commit! Generally it would be better that we direct people to make changes on `experimental/cas/main` and cherry-pick on `next` (or vice-versa), we should not be diverging.

akyrtzi replied:
I cherry-picked https://github.com/apple/llvm-project/pull/6098 to this PR to get this branch in-sync with `next` and changed `OnDiskHashMappedTrie.cpp` to guard its implementation with `LLVM_ENABLE_ONDISK_CAS` while still keeping the macro out of the header.

akyrtzi committed:
[CAS/OnDiskHashMappedTrie] Guard the implementation using `LLVM_ENABLE_ONDISK_CAS`

Guard the implementation using preprocessor guards that check `LLVM_ENABLE_ONDISK_CAS` but without putting
that macro in the `llvm/cas` header.

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan reviewed (approved):
LGTM. Thanks!

I missed the upstream review so I just merged them today.
